mirrord/agent/src/steal/connection.rs

```diff
@@ -591,28 +604,131 @@ impl TcpConnectionStealer {
+            Version::HTTP_10 => "HTTP/1.0",
+            Version::HTTP_11 => "HTTP/1.1",
+            Version::HTTP_2 => "HTTP/2",
+            Version::HTTP_3 => "HTTP/3",
+            _ => "HTTP/1.1", // TODO: What should happen here?
+        };
+
+        // Status line.
+        bytes.put(version.as_bytes()); // HTTP/1.1
+        bytes.put(&b" "[..]);
+        bytes.put(status.as_str().as_bytes()); // 200
+        bytes.put(&b" "[..]);
+        if let Some(reason) = status.canonical_reason() {
+            bytes.put(reason.as_bytes()); // OK
+        } else {
+            bytes.put(&b"LOL"[..]);
+        }
+
+        // Headers.
+        for (name, value) in headers.into_iter() {
+            if let Some(name) = name {
+                bytes.put(&b"\r\n"[..]);
+                bytes.put(name.as_str().as_bytes());
+                bytes.put(&b": "[..]);
+            } else {
+                // TODO: are multiple values for the same header comma delimited?
+                bytes.put(&b","[..])
+            }
+            bytes.put(value.as_bytes())
+        }
+        bytes.put(&b"\r\n\r\n"[..]);
+
+        // Body.
+        bytes.put(body.collect().await.unwrap().to_bytes()); // Unwrap Infallible.
+
+        bytes.to_vec()
+    }
+
+    fn handle_response_reconstruction_fail(err: hyper::http::Error) -> hyper::http::Error {
+        error!("Could not reconstruct http response: {err:?}");
+        // TODO: What should happen? Currently just skipping that response, which would
+        //       also spoils the rest of the stream. Should hopefully never happen.
+        debug_assert!(false);
+        err
+    }
+
+    /// Forward an HTTP response to a stolen HTTP request from the layer to the back to the browser.
     ///
-    ///                       _________________________agent_________________________
-    /// Local App -> Layer -> ClientConnectionHandler -> Stealer -> HttpFilterManager -> Browser
-    ///                                                          ^- You are here.
-    async fn http_response(
-        &mut self,
-        response: HttpResponse,
-    ) -> std::result::Result<(), AgentError> {
-        todo!();
-        match self.port_subscriptions.get(&response.port) {
-            Some(HttpFiltered(manager)) => todo!(),
-            Some(Unfiltered(_)) | None => {
-                warn!(
-                    "Got an http response to port {:?} which is not being filtered. \
-                Maybe the layer unsubscribed and this is a remainder of a closed connection. \
-                Not forwarding.",
-                    response.port
-                );
-                Ok(())
+    ///                         _______________agent_______________
+    /// Local App --> Layer --> ClientConnectionHandler --> Stealer --> Browser
+    ///                                                             ^- You are here.
+    async fn http_response(&mut self, response: HttpResponse) -> Result<()> {
+        let tcp_stream_write_half;
```

Comment:
Bit of a weird pattern to not initialize this directly.

I would personally go for chaining operations, but could also move everything to inside the `if let`.

Reply:
Can't `return` from the function from within closures, so I don't think we could use chained operation here.
How about this: https://github.com/t4lz/mirrord/pull/3/commits/4f631bf6b934e2162ccbcc769e24fbf40afc8b3d?

Reply:
Better, but this should be `match` instead of `if let` I think.

<- Not a fan of early returns like that though, may I try "closurefying" this function (we could also try it together)?

Reply:
Something like this:

```rust
match self
            .http_write_streams
            .get_mut(&response.connection_id)
            .and_then(|stream| {
                let counter = self
                    .http_request_counters
                    .entry(response.connection_id)
                    .or_insert(0);

                if response.request_id == *counter {
                    Some((
                        stream,
                        counter,
                        response
                            .response
                            .try_into()
                            .map_err(Self::handle_response_reconstruction_fail),
                    ))
                } else if response.request_id > *counter {
                    self.http_response_queues
                        .entry(response.connection_id)
                        .or_insert_with(|| BinaryHeap::with_capacity(8))
                        .push(response);

                    None
                } else {
                    error!(
                    "Got an http response with a request_id that was already received and sent to \
                    the browser for this connection. Dropping the duplicate response."
                    );

                    None
                }
            }) {
            Some((http_tx, counter, response)) => {
                let response = response?;
                http_tx
                    .write_all(&Self::response_to_bytes(response).await)
                    .await?; // TODO: handle error?

                *counter += 1;

                Ok(())
            }
            None => Ok(()),
            _ => Ok(()),
        }
```

Reply:
Ok, changing. Why is this better though?

Reply:
imo: Makes it easier to read code when it has a "flow", like every step does an operation, and we leave "breaks" to another place.

Like `A |> B |> C |> D` and if any step `|>` failed, then do `E`.

When you're dealing with `Option<T>` and stuff, the more you can work in the `Option` context the better, especially when using an expression based language like rust.

Reply:
This code does not do the same thing as the original code :(
Rewriting the original code to be more functional.

Reply:
@eyalb181 since this is not a merge into main, and this is a comment about style, and we're all kind of blocked on merging this code, we decided to merge this PR like this, and @meowjesty will do a separate PR for giving this function more functional style before we merge `stealer-filter` into `main`, ok?